service/history/statemachine_environment.go

```diff
@@ -234,10 +239,53 @@ func (e *stateMachineEnvironment) validateStateMachineRef(ms workflow.MutableSta
 		return fmt.Errorf("%w: %w", serviceerror.NewInternal("node lookup failed"), err)
 	}
 
+	// TODO(bergundy): When we support deletion, we'll need to also check the ref's initial version and transition count.
+
+	// Non-zero MachineLastUpdateMutableStateTransitionCount marks the reference as non-concurrent and can be
+	// invalidated with the check below.
+	if ref.StateMachineRef.MachineLastUpdateMutableStateTransitionCount == 0 {
+		return nil
+	}
+
+	if node.InternalRepr().GetLastUpdateMutableStateTransitionCount() != ref.StateMachineRef.MachineLastUpdateMutableStateTransitionCount {
+		return fmt.Errorf("%w: state machine last update transition count != ref last update transition", consts.ErrStaleReference)
+	}
+	return nil
+}
+
+func (e *stateMachineEnvironment) validateStateMachineRefWithoutTransitionHistory(ms workflow.MutableState, ref hsm.Ref, potentialStaleState bool) error {
+	// Ignore potentialStaleState if the reference can't cannot reference stale state (e.g it came from task executor
+	// not an API request).
+	potentialStaleState = potentialStaleState && ref.CanReferenceStaleState
+
+	if ms.GetCurrentVersion() < ref.StateMachineRef.MutableStateNamespaceFailoverVersion {
+		if potentialStaleState {
+			return fmt.Errorf("%w: mutable state version < ref version", consts.ErrStaleState)
+		}
+		return fmt.Errorf("%w: mutable state version < ref version", consts.ErrStaleReference)
```

Comment:
If the StateMachineRef has higher version, why would it be stale reference?

Reply:
It could be that this version isn't in our transition history but we can't really know for certain so we can only assume that the ref is stale here.

Reply:
I thought the highest failover version was always considered the "source of truth".

Reply:
Hmm.. let me double check this.